### PR TITLE
Always enforce cluster-wide shard limit

### LINKED ISSUE #20705: Hard limit on total number of shards in a cluster
We're seeing hundreds of cases of too many shards causing problems vs. problems caused by having too few.

It would be great to have a default hard limit, even though it can be increased (through the cluster settings API). It'll raise awareness to this issue hopefully in a "I can bump this now, but need to fix it"-way.

### PR DESCRIPTION
**Pull request by AthenaEryma**
This removes the option to run a cluster without enforcing the
cluster-wide shard limit, making strict enforcement the default and only
behavior.  The limit can still be adjusted as desired using the cluster
settings API.

This is a follow-up to #34021

Closes #20705

### CODE REVIEW

File: docs/reference/modules/cluster/misc.asciidoc [hunk 1]
```diff
@@ -26,44 +26,40 @@ API can make the cluster read-write again.
...
-limit will issue a deprecation warning.
-
-NOTE: You can set the system property `es.enforce_max_shards_per_node` to `true`
-to opt in to strict enforcement of the shard limit. If this system property is
-set, actions which would result in the cluster going over the limit will result
-in an error, rather than a deprecation warning. This property will be removed in
-Elasticsearch 7.0, as strict enforcement of the limit will be the default and
-only behavior.
+There is a soft limit on the number of shards in a cluster, based on the number
+of nodes in the cluster.  This is intended to prevent operations which may
```
Comment (jasontedor) [inline docs/reference/modules/cluster/misc.asciidoc]: Nit: a single space is fine; the CSS/HTML ensures consistent style anyway.

File: server/src/test/java/org/elasticsearch/cluster/shards/ClusterShardLimitIT.java [hunk 2]
```diff
@@ -55,6 +75,233 @@ public void testMinimumPerNode() {
...
+        ClusterState clusterState = client().admin().cluster().prepareState().get().getState();
+        assertFalse(clusterState.getMetaData().hasIndex("should-fail"));
+    }
+
+    public void testIncreaseReplicasOverLimit() {
+        int dataNodes = client().admin().cluster().prepareState().get().getState().getNodes().getDataNodes().size();
+
+        dataNodes = ensureMultipleDataNodes(dataNodes);
+
+        int firstShardCount = between(2,10);
```
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/cluster/shards/ClusterShardLimitIT.java]: Nit: space after the comma

File: server/src/test/java/org/elasticsearch/cluster/shards/ClusterShardLimitIT.java [hunk 3]
```diff
@@ -55,6 +75,233 @@ public void testMinimumPerNode() {
...
+
+        // Create two indexes: One that ends up with fewer shards, and one
+        // that ends up with more to verify that we check the _total_ number of
+        // shards the operation would add.
+
+        int firstIndexFactor = between (5, 10);
+        int firstIndexShards = firstIndexFactor * dataNodes;
+        int firstIndexReplicas = 0;
+
+        int secondIndexFactor = between(1,3);
```
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/cluster/shards/ClusterShardLimitIT.java]: Nit: space after comma

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationChecks.java [hunk 4]
```diff
@@ -29,7 +29,7 @@ private DeprecationChecks() {
 
     static List<Function<ClusterState, DeprecationIssue>> CLUSTER_SETTINGS_CHECKS =
         Collections.unmodifiableList(Arrays.asList(
```
Comment (jasontedor) [inline x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationChecks.java]: Collections.emptyList is unmodifiable too?

**Review by jasontedor — APPROVED** [review]
I am so sorry @gwbrown for the delay in reviewing. This is looking really great! I left a few incredibly minor nits. No need for another round.

**Comment by gwbrown** [comment]
Thanks for the review @jasontedor!

**Pull request merged**